core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2650,6 +2661,29 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def untraced[R, E, A](zio: ZIO[R, E, A]): ZIO[R, E, A] =
     zio.untraced
 
+  /**
+   * Feeds elements of type A to f and accumulates all errors in error channel or successes in success channel.
+   */
+  final def validateM[R, E, A, B](in: Iterable[A])(f: A => ZIO[R, E, B]): ZIO[R, List[E], List[B]] =
+    in.foldRight[ZIO[R, List[E], List[B]]](effectTotal(Nil)) { (a, io) =>
+      f(a).foldM(
+        e => io.mapError( _:+ e),
+        a => io.map( _ :+ a)
+      )
+    }
+
+  /**
+   * Feeds elements of type A to f until it succeeds. Returns first success or the accumulation of all errors.
+   */
+  final def validateFirstM[R, A, E, B](in: Iterable[A])(f: A => ZIO[R, E, B]): ZIO[R, List[E], B] = {
+    in.foldRight[ZIO[R, B, List[E]]](effectTotal(Nil)) { (a, io) =>
```

Comment:
Nice use of `flip`! Could implement in terms of `ZIO.foldLeft`:

```scala
ZIO.foldLeft(in)(List.empty[E]) { case (es, a) =>
  f(a).foldM(e => ZIO.succeed(e :: es), b => ZIO.fail(b))
}.flip
```

Reply:
Muuuuuch better :D Thanks